Add Executor.Priority.BACKGROUND

joshmoore opened the pull request:
The addition of parallel import has made it quite easy
to use up the omero.db.pool_size number of PostgreSQL
connections available to the server. To prevent this,
the ManagedImportRequests that are submitted on import
now go to a lower priority pool of threads.

# Testing this PR

1. with parallel-fileset, import a number of filesets concurrently
2. watch that other clients/users can still login regardless of the import progress
3. if desired, set `omero.threads.background_threads` to one and watch that import is *serial*
4. if desired, set `omero.threads.background_timeout` to something small to see client errors

joshmoore committed:
Add Executor.Priority.BACKGROUND

The addition of parallel import has made it quite easy
to use up the omero.db.pool_size number of PostgreSQL
connections available to the server. To prevent this,
the ManagedImportRequests that are submitted on import
now go to a lower priority pool of threads.

mtbc commented on components/server/src/ome/services/util/Executor.java:
```diff
@@ -354,6 +364,11 @@ public Impl(CurrentDetails principalHolder, SessionFactory factory,
             this.service = service;
             // Allowed to create more threads.
             this.systemService = Executors.newCachedThreadPool();
+            this.backgroundService = Executors.newFixedThreadPool(10,
```
Would be great if this were configurable (so we can experiment) and maybe wise to default to less than default DB pool size.

joshmoore replied:
Agreed. This is very much a temporary fix for testing purposes while we figure out the appropriate concurrency abstraction.

joshmoore committed:
Use BoundedExecutorService for background threads

Rather than define a separate pool of background threads,
a view of the USER-priority threads is provided that only
allows max_background_threads to be used. USER threads
could potentially consume all slots. If max_background_threads
is greater than max_threads, then both BACKGROUND And USER will
compete for all slots.

joshmoore commented:
The loggers which occur in the new background threads:

```
[hudson@ome-c6100-1 dist]$ grep ground var/log/Blitz-0.log  | cut -b34-71 | sort | uniq -c
      8             loci.formats.FormatHandler
     11               loci.formats.ImageReader
     44                  loci.formats.Memoizer
     20   ome.formats.enums.IQueryEnumProvider
     22      ome.formats.importer.ImportConfig
     44   ome.formats.OMEROMetadataStoreClient
     11        ome.security.basic.EventHandler
     33       ome.services.util.ServiceHandler
    187                org.perf4j.TimingLogger
     33   o.s.blitz.repo.ManagedImportRequestI
      8 o.s.b.r.ManagedImportRequestI.@448815d  (x 10)
```

joshmoore committed:
Rename/rewrite BoundedExecutor

joshmoore committed:
Replace BoundExecutor with ThreadPool itself

This turns our ThreadPool instance into a subclass of ThreadPoolExecutor
in order to control the internals of task submission. The API that is being
used is more or less intended for extension as is done here, so the risk
should be fairly minimal. Unclear is what time limit (currently 1 hour)
should be placed on blocking submission. it was also not possible to keep
the specialized background ThreadFactory implementation since the Runnable
which is passed to ThreadFactory.newThread is a ThreadPoolExecutor inner
private class.

By setting omero.threads.max_background_threads to 1, import will be restricted
to serial behavior:

```
018-08-31 10:43:00,831 INFO  [                 org.perf4j.TimingLogger] (   poola-4) start[1535704971120] time[9709] tag[omero.background.task]
2018-08-31 10:43:00,831 INFO  [                 org.perf4j.TimingLogger] (l.Server-1) start[1535704971205] time[9625] tag[omero.background.submit]
2018-08-31 10:43:03,765 INFO  [                 org.perf4j.TimingLogger] (   poola-4) start[1535704980831] time[2934] tag[omero.background.task]
2018-08-31 10:43:03,765 INFO  [                 org.perf4j.TimingLogger] (l.Server-5) start[1535704971217] time[12547] tag[omero.background.submit]
2018-08-31 10:43:06,945 INFO  [                 org.perf4j.TimingLogger] (   poola-4) start[1535704983765] time[3179] tag[omero.background.task]
2018-08-31 10:43:06,945 INFO  [                 org.perf4j.TimingLogger] (l.Server-2) start[1535704972358] time[14586] tag[omero.background.submit]
2018-08-31 10:43:10,006 INFO  [                 org.perf4j.TimingLogger] (   poola-2) start[1535704986945] time[3061] tag[omero.background.task]
2018-08-31 10:43:10,006 INFO  [                 org.perf4j.TimingLogger] (.Server-10) start[1535704972477] time[17528] tag[omero.background.submit]
2018-08-31 10:43:13,350 INFO  [                 org.perf4j.TimingLogger] (   poola-2) start[1535704990006] time[3343] tag[omero.background.task]
2018-08-31 10:43:13,350 INFO  [                 org.perf4j.TimingLogger] (l.Server-6) start[1535704972491] time[20857] tag[omero.background.submit]
2018-08-31 10:43:16,046 INFO  [                 org.perf4j.TimingLogger] (   poola-2) start[1535704993350] time[2695] tag[omero.background.task]
2018-08-31 10:43:16,046 INFO  [                 org.perf4j.TimingLogger] (l.Server-3) start[1535704972687] time[23358] tag[omero.background.submit]
2018-08-31 10:43:18,902 INFO  [                 org.perf4j.TimingLogger] (   poola-1) start[1535704996046] time[2856] tag[omero.background.task]
2018-08-31 10:43:18,902 INFO  [                 org.perf4j.TimingLogger] (l.Server-4) start[1535704972760] time[26141] tag[omero.background.submit]
2018-08-31 10:43:21,989 INFO  [                 org.perf4j.TimingLogger] (   poola-4) start[1535704998902] time[3086] tag[omero.background.task]
2018-08-31 10:43:21,989 INFO  [                 org.perf4j.TimingLogger] (.Server-11) start[1535704972818] time[29170] tag[omero.background.submit]
2018-08-31 10:43:24,986 INFO  [                 org.perf4j.TimingLogger] (   poola-3) start[1535705001989] time[2996] tag[omero.background.task]
```

Note how the wait times (omero.background.submit) increase linearly.

joshmoore commented:
Primary questions for the current implementation in this PR are:

 * Do we need to increase pool size (DB connections) by default?
 * What percentage of max_threads should background_threads be by default?
 * How long should we wait to _submit_ an import before throwing to the user?

cc: @dominikl @sbesson @mtbc @rgozim @jburel

joshmoore committed:
ThreadPool: add default constructor

joshmoore committed:
ThreadPool: add timeout configuration

joshmoore commented:
Note: configuration property changed in description

Timeout is now configurable (in ms) with `omero.threads.background_timeout`. Setting it to 1 I see:

```
/tmp/fakes/image_005.fake	2018-08-31 13:07:44,018 50017      [      main] INFO   ormats.importer.cli.LoggingImportMonitor - FILESET_UPLOAD_END
/tmp/fakes/image_005.fake	2018-08-31 13:07:44,025 50024      [      main] ERROR        ome.formats.importer.ImportLibrary - Error on import: Failed to execute 1 after ome.services.util.Executor$Impl$1@7928546fms
/tmp/fakes/image_005.fake	omero.InternalException: null
/tmp/fakes/image_005.fake		at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_25]
/tmp/fakes/image_005.fake		at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_25]
/tmp/fakes/image_005.fake		at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_25]
/tmp/fakes/image_005.fake		at java.lang.reflect.Constructor.newInstance(Constructor.java:408) ~[na:1.8.0_25]
/tmp/fakes/image_005.fake		at java.lang.Class.newInstance(Class.java:438) ~[na:1.8.0_25]
/tmp/fakes/image_005.fake		at IceInternal.BasicStream.createUserException(BasicStream.java:2779) ~[ice.jar:na]
/tmp/fakes/image_005.fake		at IceInternal.BasicStream.access$300(BasicStream.java:14) ~[ice.jar:na]
/tmp/fakes/image_005.fake		at IceInternal.BasicStream$EncapsDecoder10.throwException(BasicStream.java:3298) ~[ice.jar:na]
/tmp/fakes/image_005.fake		at IceInternal.BasicStream.throwException(BasicStream.java:2291) ~[ice.jar:na]
/tmp/fakes/image_005.fake		at IceInternal.OutgoingAsync.throwUserException(OutgoingAsync.java:399) ~[ice.jar:na]
/tmp/fakes/image_005.fake		at omero.grid.ImportProcessPrxHelper.end_verifyUpload(ImportProcessPrxHelper.java:1616) ~[blitz.jar:na]
/tmp/fakes/image_005.fake		at omero.grid.ImportProcessPrxHelper.verifyUpload(ImportProcessPrxHelper.java:1495) ~[blitz.jar:na]
/tmp/fakes/image_005.fake		at omero.grid.ImportProcessPrxHelper.verifyUpload(ImportProcessPrxHelper.java:1482) ~[blitz.jar:na]
/tmp/fakes/image_005.fake		at ome.formats.importer.ImportLibrary.importImage(ImportLibrary.java:523) ~[blitz.jar:na]
/tmp/fakes/image_005.fake		at ome.formats.importer.ImportLibrary.importCandidates(ImportLibrary.java:289) ~[blitz.jar:na]
/tmp/fakes/image_005.fake		at ome.formats.importer.cli.CommandLineImporter.start(CommandLineImporter.java:289) [blitz.jar:na]
/tmp/fakes/image_005.fake		at ome.formats.importer.cli.CommandLineImporter.main(CommandLineImporter.java:967) [blitz.jar:na]
/tmp/fakes/image_005.fake	2018-08-31 13:07:44,026 50025      [      main] INFO         ome.formats.importer.ImportLibrary - Exiting on error
/tmp/fakes/image_005.fake
/tmp/fakes/image_005.fake	==> Summary
/tmp/fakes/image_005.fake	1 file uploaded, 0 filesets created, 0 images imported, 0 errors in 0:00:26.423
```

mtbc commented on components/server/src/ome/services/scheduler/ThreadPool.java:
```diff
@@ -7,46 +7,173 @@
+                new LinkedBlockingQueue<Runnable>());
+        this.maxBackground = new Semaphore(backgroundThreads);
+        this.backgroundTimeout = backgroundTimeout;
+        log.info("ThreadPool: normal={}, background={}, timeout={}",
```
may be confused with `msTimeout`

joshmoore replied:
I _knew_ I should have taken the time to clarify that... ;)

mtbc committed:
fix minor typos in comment

mtbc commented:
Guessing that any accompanying documentation can be on top of https://github.com/openmicroscopy/ome-documentation/pull/1898.

joshmoore committed:
ThreadPool: minor logging and javadoc cleanup

mtbc commented on components/server/src/ome/services/scheduler/ThreadPool.java:
```diff
@@ -7,46 +7,173 @@
 
-    private final ExecutorService executor;
+    public ThreadPool() {
+        // Values from Executres.newCachedThreadPool
```
typo

joshmoore committed:
Fix typo

mtbc commented on components/server/src/ome/services/scheduler/ThreadPool.java:
```diff
@@ -7,46 +7,173 @@
+        try {
+            if (r instanceof BackgroundFutureTask) {
+                maxBackground.release();
+                ((BackgroundFutureTask) r).sw.stop("omero.background.task");
```
I'm trying to figure out where this stopwatch is started. :smiley:

joshmoore replied:
line 172 (https://github.com/openmicroscopy/openmicroscopy/pull/5839/files/5e501abf68bea82dde8895827a99bb44b35bd70f#diff-12c557a63afd55b5d4bb0e80d1287100R172 ) or you mean -- "documentation please" ? :smile:

mtbc replied:
Heh, no, it's fine, I just didn't realize or remember than simply creating it starts it. (Maybe need to get more sleep!)

mtbc commented on etc/omero.properties:
```diff
@@ -57,6 +57,11 @@ omero.db.authority=export.openmicroscopy.org
+# The values of omero.threads.max_threads and
+# omero.threads.background_threads play a role
+# in how many DB connections are used at any given
+# time.
```
Maybe advise to set higher than `omero.threads.background_threads` and default to such?

joshmoore replied:
Better, @mtbc?

mtbc replied:
Absolutely. :+1:

mtbc commented on components/server/src/ome/services/scheduler/ThreadPool.java:
```diff
@@ -7,46 +7,173 @@
+import org.perf4j.StopWatch;
+import org.perf4j.slf4j.Slf4JStopWatch;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
```
Don't think we use this.

joshmoore committed:
Remove unused ThreadFactoryBuilder import

joshmoore committed:
Clarify omero.properties:omero.db.poolsize docs

mtbc committed:
adjust word-wrap to increase readability

mtbc commented:
With `omero.threads.background_threads=2` and `--parallel-{upload,fileset}=3` I do see that this PR changes output like,
```
2018-09-03 13:31:38,395 10331      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_IMPORTED Step: 1 of 5  Logfile: 2739
2018-09-03 13:31:38,502 10438      [l.Client-2] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_IMPORTED Step: 1 of 5  Logfile: 2740
2018-09-03 13:31:38,685 10621      [l.Client-3] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_IMPORTED Step: 1 of 5  Logfile: 2743
2018-09-03 13:31:38,786 10722      [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 2739
2018-09-03 13:31:38,902 10838      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 2740
2018-09-03 13:31:39,019 10955      [l.Client-2] INFO   ormats.importer.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 2739
2018-09-03 13:31:39,027 10963      [l.Client-3] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 2743
2018-09-03 13:31:39,053 10989      [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_PROCESSED Step: 4 of 5  Logfile: 2739
2018-09-03 13:31:39,078 11014      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - OBJECTS_RETURNED Step: 5 of 5  Logfile: 2739
2018-09-03 13:31:39,153 11089      [l.Client-2] INFO   ormats.importer.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 2740
```
to being like,
```
2018-09-03 13:42:56,332 10062      [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 2789
2018-09-03 13:42:56,375 10105      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 2790
2018-09-03 13:42:56,392 10122      [3-thread-2] INFO   ormats.importer.cli.LoggingImportMonitor - FILE_UPLOAD_COMPLETE: /tmp/dicom/brain-8.dcm
2018-09-03 13:42:56,702 10432      [l.Client-2] INFO   ormats.importer.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 2789
2018-09-03 13:42:56,702 10432      [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 2790
2018-09-03 13:42:56,771 10501      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_PROCESSED Step: 4 of 5  Logfile: 2790
2018-09-03 13:42:56,771 10501      [l.Client-2] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_PROCESSED Step: 4 of 5  Logfile: 2789
2018-09-03 13:42:56,825 10555      [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - OBJECTS_RETURNED Step: 5 of 5  Logfile: 2790
2018-09-03 13:42:56,834 10564      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - OBJECTS_RETURNED Step: 5 of 5  Logfile: 2789
```
which makes sense.
